src/db-adapters/Mongoose/MongooseAdapter.js

```diff
@@ -603,30 +604,40 @@ export default class MongooseAdapter {
     return words.join(" ");
   }
 
-  static getIdQueryType(idOrIds) {
-    const mode = typeof idOrIds === "string" ? "findOne" : "find";
-    let idQuery;
-
-    if (typeof idOrIds === "string") {
-      if (!this.idIsValid(idOrIds)) {
-        throw new APIError(404, undefined, "No matching resource found.", "Invalid ID.");
-      }
-
-      idQuery = {_id: idOrIds};
+  static getIdQueryType(idOrIds, Model) {
+    if (idOrIds == null) {
+      return Q([ "find", undefined ]);
     }
 
-    else if (Array.isArray(idOrIds)) {
-      if (!idOrIds.every(this.idIsValid)) {
-        throw new APIError(400, undefined, "Invalid ID.");
-      }
-
-      idQuery = {_id: {"$in": idOrIds}};
-    }
+    const [ mode, ids ] = Array.isArray(idOrIds) ? [ "find", idOrIds ] : [ "findOne", [ idOrIds ] ];
+    const idValidationPromises = ids.map(id => this.idIsValid(id, Model));
 
-    return [mode, idQuery];
+    return Q.all(idValidationPromises).then((idsAreValid) => {
```

Comment:
Ahhh, right. I think I finally understand what you were saying earlier. Yes, we should be failing as soon as possible. We could maintain the semantics of `idIsValid()` by doing something like this:

```js
    const idValidationPromises = ids.map(id => this.idIsValid(id, Model).then(isValid => {
      if (!isValid) {
        throw new Error('ID is invalid');
      }
    }));
```

Do you think retaining the semantics of `idIsValid` is worth the extra complexity in `getIdQueryType`?